src/ucp/rma/flush.c

```diff
@@ -17,12 +17,22 @@
 
 static void ucp_ep_flush_error(ucp_request_t *req, ucs_status_t status)
 {
-    ucs_log_level_t level = (ucp_ep_config(req->send.ep)->key.err_mode ==
+    ucp_ep_h ep           = req->send.ep;
+    ucs_log_level_t level = (ucp_ep_config(ep)->key.err_mode ==
                              UCP_ERR_HANDLING_MODE_PEER) ?
                              UCS_LOG_LEVEL_TRACE_REQ : UCS_LOG_LEVEL_ERROR;
 
-    req->status = status;
     --req->send.state.uct_comp.count;
+    if (ucs_test_all_flags(ep->flags,
+                           UCP_EP_FLAG_FAILED |
+                           UCP_EP_FLAG_CLOSED)) {
```

Comment:
aren't we already setting "level" based on error handling mode?
or the point of the PR is avoiding "req->status = status" so close_nb request will not fail?

Reply:
> or the point of the PR is avoiding "req->status = status" so close_nb request will not fail?

yes, the point of this PR is to not set `req->status` to the error

> aren't we already setting "level" based on error handling mode?

when it fails:
```
(gdb) p req->send.ep->worker->ep_config[req->send.ep->cfg_index]->key.err_mode
$2 = UCP_ERR_HANDLING_MODE_NONE
```

the peer failure wasn't requested by the test, but UCP unconditionally provides error callback to UCT
https://github.com/openucx/ucx/blob/c197746018e102a4dc9e64c6a07a110e82f9958a/src/ucp/core/ucp_worker.c#L1197

Reply:
> the peer failure wasn't requested by the test, but UCP unconditionally provides error callback to UCT

it's ok to provide callback to UCT. as long as we don't call user callback from UCP.
In this case, the test should suppress log message. It should not expect "silent" error, when ucp ep err mode is not set.

Reply:
> it's ok to provide callback to UCT. as long as we don't call user callback from UCP.

yes, I know

> In this case, the test should suppress log message. It should not expect "silent" error, when ucp ep err mode is not set.

I guess, we should fix all tests that do disconnect
but what should we do with close request status? ignore the status?

Reply:
also, if an err handler was specified, we should fail the close request. wdyt?

Reply:
it's ok that ep_close(FLUSH) would fail if there was error during its execution
a user who doesn't care about error, would do ep_close(FORCE), or ignore error on close request

Reply:
> a user who doesn't care about error, would do ep_close(FORCE), or ignore error on close request

should we change the disconnect in the tests to ep_close(FORCE) then?

Reply:
either that, or suppress log + ignore failure of close request

Reply:
> either that

looks like, it is not... flush(FORCE) isn't supported if err handling is not requested from UCP user

Reply:
will replace the fix by suppressing the log and ignoring failure of close request

Reply:
done

Reply:
i think this is actually TCP issue. the test should not generate error at all.
it happens because sometimes the receiver does not get notification about remote connection close, so when it creates 2nd connection, it picks a "zombie" connection from conn_match

Reply:
> it happens because sometimes the receiver does not get notification about remote connection close, so when it creates 2nd connection, it picks a "zombie" connection from conn_match

it may not get a remote connection close, but since we always increment connection match ID, when we create new connection, it will try to find remote peer with new ID

Reply:
we didn't increment it on receiver side so it stayed with 0

Reply:
> we didn't increment it on receiver side so it stayed with 0

do you mean that TCP EP lives for some time on a receiver side after its connection pair was disconnected, and then it fails and calls error handling callback?